lib/hamt.js

```diff
@@ -528,31 +538,46 @@ Map.prototype.isEmpty = function() {
 /**
     Alter the value stored for `key` in `map` using function `f` using
     custom hash.
-    
+
     `f` is invoked with the current value for `k` if it exists,
     or no arguments if no such value exists. `modify` will always either
     update or insert a value into the map.
-    
+
     Returns a map with the modified value. Does not alter `map`.
 */
-const modifyHash = hamt.modifyHash = (f, hash, key, map) =>
-    new Map(map.root._modify(0, f, hash, key));
+const modifyHash = hamt.modifyHash = function(f, hash, key, defaultValue, map) {
+    if(arguments.length <= 4) {
+        map = defaultValue;
+        defaultValue = nothing;
+    }
+
+    f = defaultVal(f, defaultValue);
+
```

Comment:
This approach seems overly complicated, specifically `defaultVal`. The additional function call cost should only be paid in the case where defaultValue is provided. So move the `f = ...` into an else block.

Then, when defaultValue exists within the else block, just rewrite `defaultVal` to be something like:

```js
const defaultVal = (f, defVal) =>
     function(x) {
       return f(arguments.length === 0 ? defVal : x);
   };
```

I dont think `apply` is required. `nothing` is a purely internal concept used to implement `delete` with minimal code duplication. No external caller can use `nothing`.